Dart/src/com/jetbrains/lang/dart/ide/completion/DartServerCompletionContributor.java

```diff
@@ -99,93 +99,114 @@ protected void addCompletions(@NotNull final CompletionParameters parameters,
 
                if (file == null) return;
 
-               if (FileTypeRegistry.getInstance().isFileOfType(file, HtmlFileType.INSTANCE) &&
-                   PubspecYamlUtil.findPubspecYamlFile(project, file) == null &&
-                   !Registry.is("dart.projects.without.pubspec", false)) {
-                 return;
-               }
-
                final DartSdk sdk = DartSdk.getDartSdk(project);
                if (sdk == null || !DartAnalysisServerService.isDartSdkVersionSufficient(sdk)) return;
 
                final DartAnalysisServerService das = DartAnalysisServerService.getInstance(project);
                das.updateFilesContent();
 
                final int offset = InjectedLanguageManager.getInstance(project).injectedToHost(originalFile, parameters.getOffset());
-               final String completionId = das.completion_getSuggestions(file, offset);
-               if (completionId == null) return;
-
-               final VirtualFile targetFile = file;
-               das.addCompletions(file, completionId, (replacementOffset, replacementLength, suggestion) -> {
-                 final CompletionResultSet updatedResultSet;
-                 if (uriPrefix != null) {
-                   updatedResultSet = resultSet;
-                 }
-                 else {
-                   final String specialPrefix = getPrefixForSpecialCases(parameters, replacementOffset);
-                   if (specialPrefix != null) {
-                     updatedResultSet = resultSet.withPrefixMatcher(specialPrefix);
-                   }
-                   else {
-                     updatedResultSet = resultSet;
-                   }
+
+               if (!das.getServerVersion().isEmpty() && StringUtil.compareVersionNumbers(das.getServerVersion(), "1.33") >= 0) {
+                 final DartAnalysisServerService.CompletionInfo2 completionInfo2 = das.completion_getSuggestions2(file, offset, 100);
+                 if (completionInfo2 == null || completionInfo2.mySuggestions.isEmpty()) {
+                   return;
                  }
 
-                 LookupElementBuilder lookupElement = null;
+                 final CompletionResultSet updatedResultSet = resultSet;
```

Comment:
There are a few tricks that fix replacement offset and replacement length in the old completion provider implementation. Sometimes we calculate offsets ourselves, sometimes use info from server. If there are no known issues with the info from server, maybe let's try to always use replacement offset and replacement length from server?

Reply:
The offset and length ought to be correct as long as the completions were computed from the same state of the file as the state in which completion is being offered. But if server's view of the file (at the time suggestions are computed) can ever be different from IntelliJ's (at the time suggestions are presented to the user), then the offset and/or length could be wrong.